Health node

priyashuu opened the pull request:
###  PR: Add Node Health Dashboard with Real-Time Monitoring

#### Description:

This PR adds a **fully functional Node Health Dashboard** that streams real-time data from a Bitcoin node using WebSockets. It provides a detailed and visual view into blockchain sync status, peer connections, mempool stats, network info, and bandwidth usage.
- issue: #193 

---

### 🧪 How to Run & Test

1. **Navigate to the backend API directory:**

   ```bash
   cd dashboard/api
   ```

2. **Install dependencies:**

   ```bash
   npm install
   ```

3. **Run the WebSocket server:**

   ```bash
   node server.js
   ```

4. **Launch your frontend dev server:**

   ```bash
   npm run dev
   ```

   > You’ll now see live updates populating on the dashboard interface in real time.

---

###  Features Included
- [x]  **Sync Status Panel**: Block height, progress bar, disk usage
- [x]  **Peers Panel**: All inbound/outbound peers with IP, version, ping, it also has pagination within the Peers tab , it will show 10 peers per page 

* [x] **Network Panel**: 
  * Network version, relay fee, protocol status
  * Real-time graph using Recharts
  * Shows `totalbytesrecv` and `totalbytessent`
  * Smart unit formatting (KB, MB, GB, TB)
* [x]  **Mempool Panel**: Transaction count, memory usage, fee rate





### 💬 Open to Suggestions!

Feel free to review and suggest improvements. Whether it's on UI, performance, logic, or visual hierarchy — **open to all feedback**.

---

### 📸 Preview
<img width="1875" height="888" alt="image" src="https://github.com/user-attachments/assets/e2661b61-d362-4068-93fb-e7fd9040074a" />
<img width="1897" height="850" alt="image" src="https://github.com/user-attachments/assets/84f33048-5ca3-4fde-9492-27670678c38c" />
<img width="1854" height="777" alt="image" src="https://github.com/user-attachments/assets/13df9aa7-02ef-493b-b634-a66ddde4b8f2" />
<img width="1829" height="506" alt="image" src="https://github.com/user-attachments/assets/6714e41f-50da-4fe2-a09e-76272233cbaf" />

mcelrath committed:
doc: add contribution guidelines (#156) (#157)

Co-authored-by: Mohd Zaid <72139621+zaidmstrr@users.noreply.github.com>

u32luke committed:
Update general_considerations.md (#173)

Typo fix

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into health-node

priyashuu committed:
feat: Add Node Health Dashboard with network, mempool, and bandwidth statistics

Updated package.json to include new dependencies: cors, dotenv, express, and js-cookie.
Implemented NodeHealth component with tabs for Blockchain, Peers, Network, Mempool, and Bandwidth.
Created BandwidthPanel, MempoolPanel, NetworkPanel, and Peers components to display respective data.
Added API calls to fetch blockchain, peer, network, mempool, and net totals information.
Introduced RawJsonViewer for displaying raw data in JSON format.
Set up a proxy server to handle API requests to the Braidpool backend.

priyashuu committed:
fix: revert merge commit to prevent test failure

priyashuu committed:
feat: modularize server with routes, services, and config structure

priyashuu committed:
fix: revert merge commit to prevent test failure

priyashuu committed:
feat: add real-time last updated timestamp display

priyashuu committed:
fix: revert broken merge commit using correct parent to pass tests

priyashuu committed:
fix:UI color fixes to ensure consistency

its-me-abhishek reviewed (changes requested):
- the black highlight doesn’t work well on dark backgrounds, maybe use orange or the blue used elsewhere?
- the progress bar colors could also use better contrast.
- if Auto Refresh is on, there should be a way to turn it off.
- the Refresh button next to Auto refresh doesn’t work for me.
- also, the Raw JSON view kind of feels out of place. it reminds of Swagger UI. not sure if it fits here. maybe we can remove it ?
- api to be updated as discussed on Discord

priyashuu committed:
Remove Rawjsondata , fix progress bar colour ,

priyashuu committed:
update pages

mcelrath committed:
Merge branch 'dev' into health-node

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into health-node

priyashuu committed:
Remove deprecated API files and integrate new Node Health dashboard components

- Deleted unused API files: app.js, nodeApi.js, proxyRoute.js, and rpcClient.js.
- Enhanced server.js to include a new function for sending node health data to WebSocket clients.
- Updated NodeHealth component to utilize WebSocket for real-time data updates and added bandwidth tracking.
- Refactored Bandwidth, Mempool, and Network components for improved data handling and UI presentation.
- Introduced new types for better type safety and clarity in NodeHealth components.
- Updated Dashboard to include a new Node Health tab with relevant data visualizations.

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into health-node

priyashuu committed:
feat: add BandwidthPanel with real-time line chart and optional embedded NetworkPanel

priyashuu committed:
feat: convert total bytes sent/received to human-readable units in BandwidthPanel

priyashuu committed:
feat: add pagination to Peers panel (10 peers per page)

priyashuu committed:
Merge branch 'dev' into health-node

its-me-abhishek commented on dashboard/src/components/NodeHealth/NodeHealth.tsx:
```diff
@@ -0,0 +1,295 @@
+  );
+};
+
+const InfoRow = ({
```
we can move this into another file as a component

its-me-abhishek commented on dashboard/api/config/env.js:
```diff
@@ -0,0 +1,7 @@
+import dotenv from 'dotenv';
```
this file is not being imported anywhere, either shall use these or remove the file

its-me-abhishek commented on dashboard/api/utils/fetchBlockChainInfo.js:
```diff
@@ -0,0 +1,46 @@
+
+export async function fetchAllNodeData() {
+  try {
+    const [blockchain, peers, network, mempool, totals] = await Promise.all([
```
should have some error handling. as rpcWithEnv lacks it as well, might become hard to debug if this breaks

its-me-abhishek commented on dashboard/src/components/NodeHealth/Peers.tsx:
```diff
@@ -0,0 +1,90 @@
+
+export default function Peers({ peers }: { peers: PeerInfo[] }) {
+  const [currentPage, setCurrentPage] = useState(1);
+  const totalPages = Math.ceil(peers.length / ITEMS_PER_PAGE);
```
this can be moved to utils, or done inline if possible, as it will become easy to test

its-me-abhishek commented on dashboard/src/components/NodeHealth/Peers.tsx:
```diff
@@ -0,0 +1,90 @@
+  const [currentPage, setCurrentPage] = useState(1);
+  const totalPages = Math.ceil(peers.length / ITEMS_PER_PAGE);
+
+  const paginatedPeers = peers.slice(
```
this can be moved to utils, as well

its-me-abhishek commented on dashboard/api/utils/fetchBlockChainInfo.js:
```diff
@@ -0,0 +1,46 @@
+import { rpcWithEnv } from './rpcWithEnv.js';
+
+async function getBlockchainInfo() {
```
can these be moved to a single file? and imported everywhere from there only? here and elsewhere

its-me-abhishek replied:
Please move them to a single file for easy imports, as it will become easier to debug

its-me-abhishek commented on dashboard/src/components/NodeHealth/NodeHealth.tsx:
```diff
@@ -0,0 +1,295 @@
+        {activeTab === 'bandwidth' && (
+          <div className="space-y-4">
+            {networkInfo && <NetworkPanel network={networkInfo} />}
+            {activeTab === 'bandwidth' && (
```
this is nested already so unnecessary

its-me-abhishek commented on dashboard/src/components/NodeHealth/NodeHealth.tsx:
```diff
@@ -0,0 +1,295 @@
+    }
+  };
+
+  if (loading || !blockchainInfo) {
```
this might cause rendering even if other important data (networkInfo, mempoolInfo) is still null

its-me-abhishek commented on dashboard/src/components/NodeHealth/Bandwidth.tsx:
```diff
@@ -0,0 +1,54 @@
+const BandwidthPanel: React.FC<BandwidthPanelProps> = ({
+  bandwidthHistory,
+}) => {
+  const filteredData = bandwidthHistory.length > 1 ? bandwidthHistory : [];
```
if there's only 1 data point, why exclude it?

its-me-abhishek reviewed (changes requested):
Please fix these

its-me-abhishek commented:
Please also add unit tests for all new files, functions (API/UI) in their respective directories, and update wherever changes are made if possible.

its-me-abhishek commented:
For frontend, 

1. Empty spaces are there in many components, could be related to the margins or padding
<img width="1885" height="599" alt="Screenshot 2025-07-12 230825" src="https://github.com/user-attachments/assets/cfd176ae-e18f-43ac-9a7d-ed60ae09f6f3" />

2. Overall color of this seems a bit different from other components and border
<img width="452" height="546" alt="Screenshot 2025-07-14 224420" src="https://github.com/user-attachments/assets/fa8b9faf-d9c0-4cda-9b52-4e1a6ac3cd7c" />

its-me-abhishek commented:
Overflows are there in some components if width is reduced
<img width="624" height="185" alt="Screenshot 2025-07-14 223855" src="https://github.com/user-attachments/assets/384202aa-59ef-4d6e-b5b6-81e7e4d2db00" />
 
<img width="203" height="132" alt="Screenshot 2025-07-14 223812" src="https://github.com/user-attachments/assets/cb0abce7-7dad-4dbb-831e-5fbda63bbef5" />

The Node health button is also not visible on medium screens. 
<img width="999" height="872" alt="Screenshot 2025-07-14 221844" src="https://github.com/user-attachments/assets/3c5cf3e8-8296-4388-b3f9-2bd527735739" />

priyashuu committed:
refactor: modularize pagination logic and integrate custom theme colors into Tailwind

- Moved pagination and totalPages calculation logic to reusable utility functions
- Added fallback UI for BandwidthPanel when no data is available
- Removed redundant condition from bandwidth tab rendering

priyashuu committed:
Merge branch 'health-node' of github.com:braidpool/braidpool into health-node

priyashuu committed:
Added error handling to fetchAllNodeD,ata() .rpcwithenv for easier debugging

priyashuu committed:
test: add unit tests for Node Health tab components

priyashuu committed:
prettier formating

priyashuu committed:
Deleted unused config directory

priyashuu committed:
test: increase coverage in NodeHealth and Bandwidth components

- Added tests for NodeHealth to handle malformed WebSocket messages, closed WebSocket states, and cleanup on unmount
- Added more bandwidth history rendering test cases to BandwidthPanel
- Improved branch and line coverage across both files

its-me-abhishek commented:
<img width="1913" height="300" alt="image" src="https://github.com/user-attachments/assets/54cfce11-1fd0-4a55-8fe1-2c24a5cdfa5f" />

The hamburger icon does not work at all on medium screens.

<img width="361" height="402" alt="image" src="https://github.com/user-attachments/assets/cd7d0fcb-4f27-49b6-8b96-9044b3888327" />

Overflows on peers tab

<img width="348" height="126" alt="image" src="https://github.com/user-attachments/assets/04ed7fdb-3e49-45e5-ba44-217d67884ce0" />

Same here

its-me-abhishek commented:
<img width="1728" height="129" alt="image" src="https://github.com/user-attachments/assets/f9beccdb-cb1e-44a3-a3b2-fd61f96851c4" />

Why does this show Ping:ms? missed adding a value perhaps?

its-me-abhishek commented:
A suggestion on this graph, it currently looks as if at every timestamp, we are sending this much data, probably it is total sent and received (if im not wrong). Shall be updated accordingly as it is misleading imo

<img width="1764" height="427" alt="image" src="https://github.com/user-attachments/assets/44dcc6b8-b026-4f40-8f0f-ea46dc4f4e22" />

its-me-abhishek commented:
<img width="1868" height="543" alt="image" src="https://github.com/user-attachments/assets/e9f3bd62-b47c-48c2-be17-2896bab8f0d7" />

Empty spaces still exist, [here as suggested in 1.](https://github.com/braidpool/braidpool/pull/210/#issuecomment-3070408814), shall be fixed

its-me-abhishek commented:
<img width="1801" height="598" alt="image" src="https://github.com/user-attachments/assets/850b12d3-20f1-4fde-9603-5b1ac5c9dc54" />

The colour (diff in border, as pointed earlier) was fixed in main view, but still persists if api if off, or socket throws an error

EDIT: .. and while loading too
<img width="1800" height="520" alt="image" src="https://github.com/user-attachments/assets/d8945540-0e25-426f-bd95-4a06d2597783" />

its-me-abhishek commented:
<img width="708" height="97" alt="Screenshot 2025-07-18 185225" src="https://github.com/user-attachments/assets/df57f1be-f986-4b56-879d-c67a291cd242" />

i get this in console in NetInfo, is it related to the RPC? or the backend?

its-me-abhishek commented on dashboard/src/components/NodeHealth/NodeHealth.tsx:
```diff
@@ -0,0 +1,317 @@
+        }
+        setWsConnected(false);
+
+        if (reconnectAttempts < maxReconnectAttempts) {
```
should have cleared timeouts here as well. may cause memory leaks in some cases

its-me-abhishek replied:
unfixed, please clarify why, or any other suggestions

its-me-abhishek reviewed (changes requested):
please fix

priyashuu commented:
> <img alt="Screenshot 2025-07-18 185225" width="708" height="97" src="https://private-user-images.githubusercontent.com/114338679/468052184-df57f1be-f986-4b56-879d-c67a291cd242.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI4NTUxMzUsIm5iZiI6MTc1Mjg1NDgzNSwicGF0aCI6Ii8xMTQzMzg2NzkvNDY4MDUyMTg0LWRmNTdmMWJlLWY5ODYtNGI1Ni04NzlkLWM2N2EyOTFjZDI0Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxOFQxNjA3MTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iY2Y4YmY5ZGZkNmY2N2Q4NDgwMDJlZTBhODdiMDQzNDBkYjBhYTViNmU5NDJlM2U4ODY1YjUyMzdiOTYyODgwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Ourc6RAg9w66vS7oNHwi5ugwEB6eJXYHT6KR1g5wVzo">
> i get this in console in NetInfo, is it related to the RPC? or the backend?
Yes, it is related to RPC, (getblockchaininfo).

priyashuu committed:
refactor: update NodeHealth components for improved layout and data handling

- Refactored fetchAllNodeData to return a structured data object.
- Adjusted grid layout in Network and Peers components for better responsiveness.
- Changed background height properties in NodeHealth component for improved UI.
- Renamed 'ping' to 'pingtime' in PeerInfo type for clarity.

priyashuu committed:
fix: clear reconnect timeouts on unmount to prevent memory leaks

priyashuu committed:
fixes

its-me-abhishek commented:
```
npx tsc --noEmit
src/components/NodeHealth/__tests__/Peers.test.tsx:5:7 - error TS2322: Type '{ id: number; addr: string; version: number; subver: string; inbound: boolean; startingheight: number; synced_headers: number; synced_blocks: number; ping: number; bytessent: number; bytesrecv: number; }[]' is not assignable to type 'PeerInfo[]'.
  Property 'pingtime' is missing in type '{ id: number; addr: string; version: number; subver: string; inbound: boolean; startingheight: number; synced_headers: number; synced_blocks: number; ping: number; bytessent: number; bytesrecv: number; }' but required in type 'PeerInfo'.

5 const mockPeers: PeerInfo[] = Array.from({ length: 12 }).map((_, i) => ({
        ~~~~~~~~~

  src/components/NodeHealth/Types.ts:22:3
    22   pingtime: number;
         ~~~~~~~~
    'pingtime' is declared here.


Found 1 error in src/components/NodeHealth/__tests__/Peers.test.tsx:5
``` 
TypeScript Error in tests, Please look into it. This will cause further issues in Docker or production builds if not fixed.

Also, the refresh button does not seem to work, shall be checked...

https://github.com/user-attachments/assets/7d815953-0c35-4867-ab17-7f22b88fcae6

priyashuu commented:
> ```
> npx tsc --noEmit
> src/components/NodeHealth/__tests__/Peers.test.tsx:5:7 - error TS2322: Type '{ id: number; addr: string; version: number; subver: string; inbound: boolean; startingheight: number; synced_headers: number; synced_blocks: number; ping: number; bytessent: number; bytesrecv: number; }[]' is not assignable to type 'PeerInfo[]'.
>   Property 'pingtime' is missing in type '{ id: number; addr: string; version: number; subver: string; inbound: boolean; startingheight: number; synced_headers: number; synced_blocks: number; ping: number; bytessent: number; bytesrecv: number; }' but required in type 'PeerInfo'.
> 
> 5 const mockPeers: PeerInfo[] = Array.from({ length: 12 }).map((_, i) => ({
>         ~~~~~~~~~
> 
>   src/components/NodeHealth/Types.ts:22:3
>     22   pingtime: number;
>          ~~~~~~~~
>     'pingtime' is declared here.
> 
> 
> Found 1 error in src/components/NodeHealth/__tests__/Peers.test.tsx:5
> ```
> 
> TypeScript Error in tests, Please look into it. This will cause further issues in Docker or production builds if not fixed.
> 
> Also, the refresh button does not seem to work, shall be checked...
> 
>  refresh.does.not.work.mp4

It is fixed now, I already pushed the changes as I mentioned on Discord. Did you pull the latest changes?

priyashuu committed:
refactor: enhance BandwidthPanel and NodeHealth data handling

- Updated BandwidthPanel to display real-time bandwidth usage in Bytes/sec.
- Modified NodeHealth component to calculate bandwidth received and sent per second.
- Removed manual refresh button and associated logic for cleaner UI.
- Improved data structure for bandwidth history to include calculated bandwidth values.

priyashuu committed:
prettier format

its-me-abhishek commented on dashboard/src/components/NodeHealth/Bandwidth.tsx:
```diff
@@ -0,0 +1,72 @@
+  return (
+    <div className="bg-[#1e1e1e] border border-gray-700 rounded-xl shadow-md p-4">
+      <h3 className="text-lg font-semibold text-white mb-4 text-center">
+        Real-Time Bandwidth Usage (Bytes/sec)
```
is Real-Time Bandwidth Usage (Bytes/sec) the right title if data  if the graph updates every 10 sec?

priyashuu replied:
Removed the (bytes/sec) label as it's tied to the server-side update frequency (currently every 10 seconds).

Thanks for pointing it out! I’ve now ensured that the previous timeout is cleared before setting a new one in the onclose handler to avoid any potential memory leaks.

its-me-abhishek reviewed (changes requested):
PLEASE FIX

its-me-abhishek commented:
<img width="425" height="678" alt="image" src="https://github.com/user-attachments/assets/ae7babe4-8743-48b8-856b-95bb46da6339" />

Overflow on peers tab, also alignment is sort of required, as it is hard to interpret as of now

<img width="1718" height="708" alt="image" src="https://github.com/user-attachments/assets/585588f5-1bb7-42b0-b256-2c2c6a4c8280" />

Please add some x empty cards if numberofpeers%10 != 0, as it provides a smoother experience if user wants to go back

priyashuu commented:
> PLEASE FIX



> <img alt="image" width="425" height="678" src="https://private-user-images.githubusercontent.com/114338679/468395597-ae7babe4-8743-48b8-856b-95bb46da6339.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMwMjgwNDksIm5iZiI6MTc1MzAyNzc0OSwicGF0aCI6Ii8xMTQzMzg2NzkvNDY4Mzk1NTk3LWFlN2JhYmU0LTg3NDMtNDhiOC04NTZiLTk1YmI0NmRhNjMzOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyMFQxNjA5MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NzgxOTFiNDBkYmI1MjViMWIwZTlhMzBjNDMzNzI2MWIyZmZjZjIxYWMwMWRmNzE4Mzg1ZTQ2NzU4ZDUyODUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.62EuJCdDuRLLLoYffh4nmEJ6JjIBi7hKl7IHOYXRlFY">
> Overflow on peers tab, also alignment is sort of required, as it is hard to interpret as of now
> 
> <img alt="image" width="1718" height="708" src="https://private-user-images.githubusercontent.com/114338679/468395747-585588f5-1bb7-42b0-b256-2c2c6a4c8280.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMwMjgwNDksIm5iZiI6MTc1MzAyNzc0OSwicGF0aCI6Ii8xMTQzMzg2NzkvNDY4Mzk1NzQ3LTU4NTU4OGY1LTFiYjctNDJiMC1iMjU2LTJjMmM2YTRjODI4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyMFQxNjA5MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZGJiOThhYmQwZWE4MjAzYzNmZDZjMzdjY2UxNzkwMWM3NGEwZGE0NDU3YTc0YmFmMTk4NWQ5YjA0N2U2NDQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Cp1JaApLUmOf9t_zNBEuyhn6TTs_DlCO-ozfa9xO0bQ">
> Please add some x empty cards if numberofpeers%10 != 0, as it provides a smoother experience if user wants to go back
<img width="205" height="656" alt="image" src="https://github.com/user-attachments/assets/91856499-54b9-4a3d-a4e3-8b4494b5efc1" />

<img width="1847" height="692" alt="image" src="https://github.com/user-attachments/assets/5df302c3-7990-41a1-b6f9-dddcd62e83aa" />

priyashuu commented:
> PLEASE FIX

fixed

priyashuu committed:
fix: clear reconnect timeout on WebSocket close to prevent memory leaks

- Ensures previous timeout is cleared before setting a new one in `onclose`
- Removed "(bytes/sec)" label as update interval is defined by the server (currently 10s)

priyashuu committed:
fixed tabs overflow

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into health-node

its-me-abhishek commented:
1. The alignment of this can certainly be improved, as there's a lot of empty space, and currently it looks a bit messy 

<img width="205" height="656" alt="image" src="https://github.com/user-attachments/assets/a4a6a25b-a7ae-455d-8289-0f620a5d537c" />


https://github.com/user-attachments/assets/69868f4d-84d4-4cc6-aad3-46ccdae4f8d4


2. The empty cards are not working as expected, they are not  == 10, still making jitters when someone goes prev and next.

priyashuu commented:
<img width="439" height="714" alt="image" src="https://github.com/user-attachments/assets/d53091df-69a9-400d-8ec8-50ebd395dc9a" />

priyashuu committed:
fix the alignment of peers tabs ,prevent jitter effect . updated 5 cards per page

priyashuu committed:
prettier formatting

priyashuu committed:
Fixed the height of the peer cards on small screens as well

priyashuu committed:
Fix: Adjusted Peer card layout and height responsiveness for small screens
- Ensured consistent height on mobile view
- Improved flex behavior and spacing

priyashuu committed:
prettier formatting

its-me-abhishek reviewed (approved):
Tests need to be fixed and added
